pages/gaming/[id].js

```diff
@@ -0,0 +1,460 @@
+import { useRouter } from "next/router";
+import styled from "styled-components";
+import RegularButton from "@/components/Buttons/RegularButton";
+import RoundButton from "@/components/Buttons/RoundButton";
+import MarkAsIncorrect from "@/public/icons/MarkAsIncorrect.svg";
+import ButtonWrapper from "@/components/Buttons/ButtonWrapper";
+import { useState, useEffect, useMemo, useCallback } from "react";
+import { uid } from "uid";
+
+export default function TrainingCollectionPage({
+  getAllFlashcardsFromCollection,
+}) {
+  const router = useRouter();
+  const { id } = router.query;
+
+  const [isCancel, setIsCancel] = useState(false);
+  const [flippedCards, setFlippedCards] = useState([]);
+  const [memoryCards, setMemoryCards] = useState([]);
+  const [selectedCardId, setSelectedCardId] = useState(null);
+  const [cardStatus, setCardStatus] = useState({});
+  const [showOverlay, setShowOverlay] = useState(false);
+  const [isGameWon, setIsGameWon] = useState(false);
```

Comment:
I've been thinking about this for a while. 🤯 This morning I wrote a function `checkWinning()` that is executed on click when the enlarged view of the flashcard is closed. I've just started to rebuild the code and used the exact query that I had in the `checkWinning`-function now in the return statement as conditional rendering. So, specifically, I no longer wrote
```
{!isGameWon && (...)}
{isGameWon && (...)}
```
but instead wrote
```
{(memoryCards.every(
(card) =>
cardStatus[card.id] === "hidden" || flippedCards.includes(card.id)
)) ? (...)
: (...) }
```

But on closer inspection, I found the solution more confusing, so I've undone it and would like to keep the useState. 😅🙈

I try to explain my reasons:

➡️ 1. The useState `isGameWon` makes the code easier to understand in my opinion. Firstly, I have the "normal" return first and the exception (game won) after that, which I find clearer, and secondly, the expression `isGameWon` is understandable for both me and the others at first glance, whereas with the other condition it may not be clear at first glance what it is all about.

➡️ 2. In addition, the `checkWinning`-function is called quite frequently (every time the enlarged view is closed), but at least not every time the code is rendered, which is probably an advantage for performance, right? Especially since with the detailed condition the render cycle would be triggered again and again, as the *status* of the cards is constantly changing. Then the page would have even more render repetitions. Or am I making a mistake in my thinking?

I see several advantages in the useState and would actually be reluctant to delete it. But maybe I'm making a mistake in my thinking or am overlooking something? 🤔

💫 When it comes to the fact that there are too many useStates, as you suggested, I have another idea:

I have two Booleans as useStates, namely `isCancel` and `isGameWon`. Couldn't I combine the two in one useState and use strings instead, similar to what we did with `actionMode`? I think I could, for example, implement a useState `gameStatus` (naming like `cardStatus`) and then give strings, e.g. `playing` (as default), `cancel` or `won`. That looks like the best solution to me because I save a useState, but at the same time the readability of the code is not affected. Such a useState would still be clearly understandable (even in the return statement, e.g. `{(gameStatus === "playing") && (...)}` I find easy to understand). We could even add new states. I definitely won't do that soon because we're implementing the backend first, but it seems like a flexible solution to me. 💪

What do you think?

Thanks for thinking along and all your help! 💕👋

Reply:
I just changed a few things (especially improved the responsive design) and at the same time adjusted the useStates as I intended. I just commented out the old code in case you don't think it's a good idea to organize the useState the way I just changed it. If you have time, it would be great if you could take a look and tell me what you think. :blush:

Reply:
I like your idea with the `gameStatus`.

You can actually eliminate the `isGameWon` state with a little change:
You can use the same logic to check if the game is won but instead of saving it in a state, you can simply save it in a variable. 🙂 You could do that right before the return statement.

```js
  const isGameWon = memoryCards.every(
    (card) => cardStatus[card.id] === "hidden" || flippedCards.includes(card.id)
  );
```

It's up to you if you want to stick with your `gameStatus `state (as you integrate the `isCancel` state inside) or if you decide to implement my suggestion. 🙂

Reply:
Thank you! I'm just going to leave it as it is (after speaking with Phillip). I think it's a clear solution, and I think it could be expanded quite flexibly in the future (and to be honest I don't feel like rebuilding it again). :wink:
Thank you for your time and for thinking along with me! :pray: